rust/src/ike/detect.rs

```diff
@@ -0,0 +1,227 @@
+
+    if tx.ike_version == 1 {
+        if let Some(r) = tx.hdr.ikev1_header.exchange_type {
+            unsafe {
+                *exch_type = r;
+            }
+            return 1;
+        }
+    } else if tx.ike_version == 2 {
+        unsafe {
+            *exch_type = tx.hdr.ikev2_header.exch_type.0;
+        }
+        return 1;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_ike_state_get_spi_initiator(
+    tx: &mut IKETransaction, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> u8 {
+    debug_validate_bug_on!(buffer == std::ptr::null_mut() || buffer_len == std::ptr::null_mut());
+
+    unsafe {
+        *buffer = tx.hdr.spi_initiator.as_ptr();
+        *buffer_len = tx.hdr.spi_initiator.len() as u32;
+    }
+    return 1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_ike_state_get_spi_responder(
+    tx: &mut IKETransaction, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> u8 {
+    debug_validate_bug_on!(buffer == std::ptr::null_mut() || buffer_len == std::ptr::null_mut());
+
+    unsafe {
+        *buffer = tx.hdr.spi_responder.as_ptr();
+        *buffer_len = tx.hdr.spi_responder.len() as u32;
+    }
+    return 1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_ike_state_get_nonce(
+    tx: &mut IKETransaction, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> u8 {
+    debug_validate_bug_on!(buffer == std::ptr::null_mut() || buffer_len == std::ptr::null_mut());
+
+    if tx.ike_version == 1 && !tx.hdr.ikev1_header.nonce.is_empty() {
+        let p = &tx.hdr.ikev1_header.nonce;
+        unsafe {
+            *buffer = p.as_ptr();
+            *buffer_len = p.len() as u32;
+        }
+        return 1;
+    }
+
+    unsafe {
+        *buffer = ptr::null();
+        *buffer_len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_ike_state_get_key_exchange(
+    tx: &mut IKETransaction, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> u8 {
+    debug_validate_bug_on!(buffer == std::ptr::null_mut() || buffer_len == std::ptr::null_mut());
+
+    if tx.ike_version == 1 && !tx.hdr.ikev1_header.key_exchange.is_empty() {
+        let p = &tx.hdr.ikev1_header.key_exchange;
+        unsafe {
+            *buffer = p.as_ptr();
+            *buffer_len = p.len() as u32;
+        }
+        return 1;
+    }
+
+    unsafe {
+        *buffer = ptr::null();
+        *buffer_len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_ike_tx_get_vendor(
+    tx: &IKETransaction, i: u16, buf: *mut *const u8, len: *mut u32,
+) -> u8 {
+    if tx.ike_version == 1 && i < tx.hdr.ikev1_header.vendor_ids.len() as u16 {
+        unsafe {
+            *len = tx.hdr.ikev1_header.vendor_ids[i as usize].len() as u32;
+            *buf = tx.hdr.ikev1_header.vendor_ids[i as usize].as_ptr();
+        }
+        return 1;
+    }
+
+    unsafe {
+        *buf = ptr::null();
+        *len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_ike_state_get_sa_attribute(
```

Comment:
for an unsafe function, this is a bit long. Can we split the meat of the function into a safe function?

Reply:
I think so. I was able to narrow this down to just two unsafe lines of code.